etcd-recovery/etcdrecovery.go

```diff
@@ -0,0 +1,538 @@
+	ctrl.SetLogger(zap.New(zap.JSONEncoder(func(o *zapcore.EncoderConfig) {
+		o.EncodeTime = zapcore.RFC3339TimeEncoder
+	})))
+	ctx := ctrl.LoggerInto(cmdContext, ctrl.Log.WithName("etcd-recovery"))
+	return ctx
+}
+
+type etcdStatus struct {
+	failingPod    *string
+	missingMember *string
+	recoverable   bool
+}
+
+func (s *etcdStatus) isUnhealthy() bool {
+	return s.missingMember != nil || s.failingPod != nil
+}
+
+func (s *etcdStatus) isRecoverable() bool {
+	recoverable := s.recoverable
+	if s.failingPod != nil && s.missingMember != nil {
+		if *s.failingPod != *s.missingMember {
+			recoverable = false
+		}
+	}
+	return recoverable
+}
+
+func (s *etcdStatus) memberToRecover() string {
+	if !s.isRecoverable() {
+		return ""
+	}
+	if s.failingPod != nil {
+		return *s.failingPod
+	}
+	if s.missingMember != nil {
+		return *s.missingMember
+	}
+	return ""
+}
+
+func (s *etcdStatus) String() string {
+	var status []string
+	if s.missingMember != nil {
+		status = append(status, fmt.Sprintf("missing member: %s", *s.missingMember))
+	}
+	if s.failingPod != nil {
+		status = append(status, fmt.Sprintf("failing pod: %s", *s.failingPod))
+	}
+	if !s.recoverable {
+		status = append(status, "broken: NOT RECOVERABLE")
+	}
+	if len(status) > 0 {
+		return strings.Join(status, ",")
+	}
+	return "healthy"
+}
+
+func NewStatusCommand(opts *options) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:          "status",
+		Short:        "Determine if etcd cluster is healthy",
+		SilenceUsage: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx, cancel := signal.NotifyContext(cmd.Context(), os.Interrupt, syscall.SIGINT)
+			defer cancel()
+			ctx = setupCmdContext(ctx)
+			log := ctrl.LoggerFrom(ctx)
+			etcdStatus, err := runStatus(ctx, *opts)
+			if err != nil {
+				log.Error(err, "Error occurred")
+				os.Exit(1)
+			}
+			log.Info("etcd status", "status", etcdStatus.String())
+		},
+	}
+
+	return cmd
+}
+
+func runRecovery(ctx context.Context, opts options) error {
+	status, err := runStatus(ctx, opts)
+	if err != nil {
+		return err
+	}
+	if !status.isRecoverable() {
+		return errors.New("etcd cluster does not have a recoverable status")
+	}
+	if !status.isUnhealthy() {
+		return nil
+	}
+
+	if err := recoverEtcd(ctx, opts, *status); err != nil {
+		return err
+	}
+
+	if err := waitForHealthyCluster(ctx, opts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func waitForHealthyCluster(ctx context.Context, opts options) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Waiting for etcd cluster to become healthy")
+	timeOutCtx, cancel := context.WithTimeout(ctx, healthyClusterWait)
+	defer cancel()
+	return wait.PollUntilContextCancel(timeOutCtx, 30*time.Second, false, func(ctx context.Context) (done bool, err error) {
+		status, err := runStatus(ctx, opts)
+		if err != nil {
+			return false, err
+		}
+		isHealthy := !status.isUnhealthy()
+		if isHealthy {
+			log.Info("Cluster is now healthy, finished recovery job")
```

Comment:
really nit pick but logging "finished recovery job" seems out of scope of this func. It is making a non guaranteed assumption about the invoker